File: sdk/golang/tests/e2e/rest/broker_test/ndbroker_test.go
package broker_test

import (
	"context"
	"encoding/json"
	"fmt"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/extension/interceptor"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/api"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/common/logger"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/broker/ndbroker"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/types"
	"github.com/google/uuid"
	"os"
	"testing"
)

var ndbrokerApi ndbroker.NDBrokerAPI

func init() {
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	brokerName := os.Getenv("BROKER_NAME")
	brokerKey := os.Getenv("BROKER_KEY")
	brokerPartner := os.Getenv("BROKER_PARTNER")

	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	httpOptionBuilder := types.NewTransportOptionBuilder().
		SetKeepAlive(true).
		AddInterceptors(interceptor.NewLoggingInterceptor(false, defaultLogger))

	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithBrokerName(brokerName).
		WithBrokerPartner(brokerPartner).
		WithBrokerKey(brokerKey).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithTransportOption(httpOptionBuilder.Build()).
		Build()

	client := api.NewClient(option)

	ndbrokerApi = client.RestService().GetBrokerService().GetNDBrokerAPI()
}

func TestNDBrokerGetDepositListReq(t *testing.T) {
	// GetDepositList
	// Get Deposit List
	// /api/v1/asset/ndbroker/deposit/list

	builder := ndbroker.NewGetDepositListReqBuilder()
	builder.SetCurrency("USDT")
	req := builder.Build()

	resp, err := ndbrokerApi.GetDepositList(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestNDBrokerDeleteSubAccountAPIReq(t *testing.T) {
	// DeleteSubAccountAPI
	// Delete SubAccount API
	// /api/v1/broker/nd/account/apikey

	builder := ndbroker.NewDeleteSubAccountAPIReqBuilder()
	builder.SetUid("229317507").SetApiKey("67447006890ba200018d2722")
	req := builder.Build()

	resp, err := ndbrokerApi.DeleteSubAccountAPI(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}
func TestNDBrokerGetSubAccountAPIReq(t *testing.T) {
	// GetSubAccountAPI
	// Get SubAccount API
	// /api/v1/broker/nd/account/apikey

	builder := ndbroker.NewGetSubAccountAPIReqBuilder()
	builder.SetUid("226383154").SetApiKey("67446f679d8f66000107eb6d")
	req := builder.Build()

	resp, err := ndbrokerApi.GetSubAccountAPI(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}
func TestNDBrokerAddSubAccountApiReq(t *testing.T) {
	// AddSubAccountApi
	// Add SubAccount API
	// /api/v1/broker/nd/account/apikey

	builder := ndbroker.NewAddSubAccountApiReqBuilder()
	builder.SetUid("229317507").SetPassphrase("11223344").
		SetIpWhitelist([]string{"127.0.0.1", "192.168.1.1"}).SetPermissions([]string{"general", "spot"}).SetLabel("This is remarks")
	req := builder.Build()

	resp, err := ndbrokerApi.AddSubAccountApi(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestNDBrokerGetSubAccountReq(t *testing.T) {
	// GetSubAccount
	// Get SubAccount
	// /api/v1/broker/nd/account

	builder := ndbroker.NewGetSubAccountReqBuilder()
	builder.SetUid("229317507")
	req := builder.Build()

	resp, err := ndbrokerApi.GetSubAccount(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}
func TestNDBrokerAddSubAccountReq(t *testing.T) {
	// AddSubAccount
	// Add SubAccount
	// /api/v1/broker/nd/account

	builder := ndbroker.NewAddSubAccountReqBuilder()
	builder.SetAccountName("sdk_test")
	req := builder.Build()

	resp, err := ndbrokerApi.AddSubAccount(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestNDBrokerModifySubAccountApiReq(t *testing.T) {
	// ModifySubAccountApi
	// Modify SubAccount API
	// /api/v1/broker/nd/account/update-apikey

	builder := ndbroker.NewModifySubAccountApiReqBuilder()
	builder.SetUid("229317507").SetApiKey("67447006890ba200018d2722").SetIpWhitelist([]string{"127.0.0.1"}).
		SetPermissions([]string{"general"}).SetLabel("This is remarks")
	req := builder.Build()

	resp, err := ndbrokerApi.ModifySubAccountApi(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestNDBrokerGetBrokerInfoReq(t *testing.T) {
	// GetBrokerInfo
	// Get Broker Info
	// /api/v1/broker/nd/info

	builder := ndbroker.NewGetBrokerInfoReqBuilder()
	builder.SetTradeType("1")
	req := builder.Build()

	resp, err := ndbrokerApi.GetBrokerInfo(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestNDBrokerGetRebaseReq(t *testing.T) {
	// GetRebase
	// Get Broker Rebate
	// /api/v1/broker/nd/rebase/download

	builder := ndbroker.NewGetRebaseReqBuilder()
	builder.SetBegin("20240610").SetEnd("20241010").SetTradeType("1")
	req := builder.Build()

	resp, err := ndbrokerApi.GetRebase(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestNDBrokerTransferReq(t *testing.T) {
	// Transfer
	// Transfer
	// /api/v1/broker/nd/transfer

	builder := ndbroker.NewTransferReqBuilder()
	builder.SetCurrency("USDT").SetAmount("0.01").SetDirection("OUT").SetAccountType("TRADE").
		SetSpecialUid("229317507").SetSpecialAccountType("MAIN").SetClientOid(uuid.NewString())
	req := builder.Build()

	resp, err := ndbrokerApi.Transfer(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestNDBrokerGetDepositDetailReq(t *testing.T) {
	// GetDepositDetail
	// Get Deposit Detail
	// /api/v3/broker/nd/deposit/detail

	builder := ndbroker.NewGetDepositDetailReqBuilder()
	builder.SetCurrency("USDT").SetHash("6724e363a492800007ec602b")
	req := builder.Build()

	resp, err := ndbrokerApi.GetDepositDetail(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestNDBrokerGetTransferHistoryReq(t *testing.T) {
	// GetTransferHistory
	// Get Transfer History
	// /api/v3/broker/nd/transfer/detail

	builder := ndbroker.NewGetTransferHistoryReqBuilder()
	builder.SetOrderId("671b4600c1e3dd000726866d")
	req := builder.Build()

	resp, err := ndbrokerApi.GetTransferHistory(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestNDBrokerGetWithdrawDetailReq(t *testing.T) {
	// GetWithdrawDetail
	// Get Withdraw Detail
	// /api/v3/broker/nd/withdraw/detail

	builder := ndbroker.NewGetWithdrawDetailReqBuilder()
	builder.SetWithdrawalId("674686fa1ac01f0007b25768")
	req := builder.Build()

	resp, err := ndbrokerApi.GetWithdrawDetail(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

File: sdk/python/test/e2e/rest/broker_test/ndbroker_test.py
import os
import unittest
import uuid

from kucoin_universal_sdk.api.client import DefaultClient
from kucoin_universal_sdk.extension.interceptor.logging import LoggingInterceptor
from kucoin_universal_sdk.generate.broker.ndbroker.model_add_sub_account_api_req import AddSubAccountApiReqBuilder, \
    AddSubAccountApiReq
from kucoin_universal_sdk.generate.broker.ndbroker.model_add_sub_account_req import AddSubAccountReqBuilder
from kucoin_universal_sdk.generate.broker.ndbroker.model_delete_sub_account_api_req import DeleteSubAccountApiReqBuilder
from kucoin_universal_sdk.generate.broker.ndbroker.model_get_broker_info_req import GetBrokerInfoReqBuilder, \
    GetBrokerInfoReq
from kucoin_universal_sdk.generate.broker.ndbroker.model_get_deposit_detail_req import GetDepositDetailReqBuilder
from kucoin_universal_sdk.generate.broker.ndbroker.model_get_deposit_list_req import GetDepositListReqBuilder
from kucoin_universal_sdk.generate.broker.ndbroker.model_get_rebase_req import GetRebaseReqBuilder, GetRebaseReq
from kucoin_universal_sdk.generate.broker.ndbroker.model_get_sub_account_api_req import GetSubAccountApiReqBuilder
from kucoin_universal_sdk.generate.broker.ndbroker.model_get_sub_account_req import GetSubAccountReqBuilder
from kucoin_universal_sdk.generate.broker.ndbroker.model_get_transfer_history_req import GetTransferHistoryReqBuilder
from kucoin_universal_sdk.generate.broker.ndbroker.model_get_withdraw_detail_req import GetWithdrawDetailReqBuilder
from kucoin_universal_sdk.generate.broker.ndbroker.model_modify_sub_account_api_req import \
    ModifySubAccountApiReqBuilder, ModifySubAccountApiReq
from kucoin_universal_sdk.generate.broker.ndbroker.model_transfer_req import TransferReqBuilder, TransferReq
from kucoin_universal_sdk.model.client_option import ClientOptionBuilder
from kucoin_universal_sdk.model.constants import GLOBAL_API_ENDPOINT, GLOBAL_BROKER_API_ENDPOINT, \
    GLOBAL_FUTURES_API_ENDPOINT
from kucoin_universal_sdk.model.transport_option import TransportOptionBuilder


class BrokerApiTest(unittest.TestCase):
    def setUp(self):
        key = os.getenv("API_KEY")
        secret = os.getenv("API_SECRET")
        passphrase = os.getenv("API_PASSPHRASE")
        broker_name = os.getenv("BROKER_NAME")
        broker_key = os.getenv("BROKER_KEY")
        broker_partner = os.getenv("BROKER_PARTNER")

        # create http transport option
        http_transport_option = (
            TransportOptionBuilder()
            .set_interceptors([LoggingInterceptor()])
            .build()
        )

        # create client option
        client_option = (
            ClientOptionBuilder()
            .set_key(key)
            .set_secret(secret)
            .set_passphrase(passphrase)
            .set_spot_endpoint(GLOBAL_API_ENDPOINT)
            .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
            .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
            .set_broker_name(broker_name)
            .set_broker_partner(broker_partner)
            .set_broker_key(broker_key)
            .set_transport_option(http_transport_option)
            .build()
        )

        # create API client
        client = DefaultClient(client_option)
        kucoin_rest_api = client.rest_service()
        self.api = kucoin_rest_api.get_broker_service().get_nd_broker_api()

    def test_get_deposit_list_req(self):
        """
            get_deposit_list
            Get Deposit List
            /api/v1/asset/ndbroker/deposit/list
        """

        builder = GetDepositListReqBuilder()
        builder.set_currency("USDT")
        req = builder.build()
        try:
            resp = self.api.get_deposit_list(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_delete_sub_account_api_req(self):
        """
            delete_sub_account_api
            Delete SubAccount API
            /api/v1/broker/nd/account/apikey
        """

        builder = DeleteSubAccountApiReqBuilder()
        builder.set_uid("229317507").set_api_key("674ff687ad1e6600015aa2b2")
        req = builder.build()
        try:
            resp = self.api.delete_sub_account_api(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_get_sub_account_api_req(self):
        """
            get_sub_account_api
            Get SubAccount API
            /api/v1/broker/nd/account/apikey
        """

        builder = GetSubAccountApiReqBuilder()
        builder.set_uid("229317507").set_api_key("674ff687ad1e6600015aa2b2")
        req = builder.build()
        try:
            resp = self.api.get_sub_account_api(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_add_sub_account_api_req(self):
        """
            add_sub_account_api
            Add SubAccount API
            /api/v1/broker/nd/account/apikey
        """

        builder = AddSubAccountApiReqBuilder()
        builder.set_uid("229317507").set_passphrase("11223344").set_ip_whitelist(
            ["127.0.0.1", "192.168.1.1"]).set_permissions(
            [AddSubAccountApiReq.PermissionsEnum.GENERAL, AddSubAccountApiReq.PermissionsEnum.SPOT]).set_label(
            "This is remarks")
        req = builder.build()
        try:
            resp = self.api.add_sub_account_api(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_get_sub_account_req(self):
        """
            get_sub_account
            Get SubAccount
            /api/v1/broker/nd/account
        """

        builder = GetSubAccountReqBuilder()
        builder.set_uid("229317507").set_current_page(1).set_page_size(10)
        req = builder.build()
        try:
            resp = self.api.get_sub_account(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_add_sub_account_req(self):
        """
            add_sub_account
            Add SubAccount
            /api/v1/broker/nd/account
        """

        builder = AddSubAccountReqBuilder()
        builder.set_account_name("sdk_test_2")
        req = builder.build()
        try:
            resp = self.api.add_sub_account(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_modify_sub_account_api_req(self):
        """
            modify_sub_account_api
            Modify SubAccount API
            /api/v1/broker/nd/account/update-apikey
        """

        builder = ModifySubAccountApiReqBuilder()
        builder.set_uid("229317507").set_ip_whitelist(["127.0.0.1"]).set_permissions(
            [ModifySubAccountApiReq.PermissionsEnum.SPOT]).set_label("label").set_api_key("674ff687ad1e6600015aa2b2")
        req = builder.build()
        try:
            resp = self.api.modify_sub_account_api(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_get_broker_info_req(self):
        """
            get_broker_info
            Get Broker Info
            /api/v1/broker/nd/info
        """

        builder = GetBrokerInfoReqBuilder()
        builder.set_trade_type(GetBrokerInfoReq.TradeTypeEnum.T_1)
        req = builder.build()
        try:
            resp = self.api.get_broker_info(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_get_rebase_req(self):
        """
            get_rebase
            Get Broker Rebate
            /api/v1/broker/nd/rebase/download
        """

        builder = GetRebaseReqBuilder()
        builder.set_begin("20240610").set_end("20241010").set_trade_type(GetRebaseReq.TradeTypeEnum.T_1)
        req = builder.build()
        try:
            resp = self.api.get_rebase(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_transfer_req(self):
        """
            transfer
            Transfer
            /api/v1/broker/nd/transfer
        """

        builder = TransferReqBuilder()
        builder.set_currency("USDT").set_amount("0.01").set_direction(
            TransferReq.DirectionEnum.OUT).set_account_type(TransferReq.AccountTypeEnum.TRADE).set_special_uid(
            "229317507").set_special_account_type(TransferReq.SpecialAccountTypeEnum.MAIN).set_client_oid(
            uuid.uuid4().__str__())
        req = builder.build()
        try:
            resp = self.api.transfer(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_get_deposit_detail_req(self):
        """
            get_deposit_detail
            Get Deposit Detail
            /api/v3/broker/nd/deposit/detail
        """

        builder = GetDepositDetailReqBuilder()
        builder.set_currency("USDT").set_hash("6724e363a492800007ec602b")
        req = builder.build()
        try:
            resp = self.api.get_deposit_detail(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_get_transfer_history_req(self):
        """
            get_transfer_history
            Get Transfer History
            /api/v3/broker/nd/transfer/detail
        """

        builder = GetTransferHistoryReqBuilder()
        builder.set_order_id("671b4600c1e3dd000726866d")
        req = builder.build()
        try:
            resp = self.api.get_transfer_history(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_get_withdraw_detail_req(self):
        """
            get_withdraw_detail
            Get Withdraw Detail
            /api/v3/broker/nd/withdraw/detail
        """

        builder = GetWithdrawDetailReqBuilder()
        builder.set_withdrawal_id("674686fa1ac01f0007b25768")
        req = builder.build()
        try:
            resp = self.api.get_withdraw_detail(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e
